### PR TITLE
Yeyu/hf eagle medusa

### PR DESCRIPTION
**Pull request by yeyu-nvidia**
## What does this PR do?
new feature,

**Overview:** 
This PR implements HF parallel draft by combining eagle and medusa. In training, multiple medusa heads are added and trained together with eagle. In inference, medusa heads are used to generate draft tokens after all eagle tokens.

## Usage
Set parallel_draft_step>1 in eagle_config to enable parallel draft.

```python
# Add a code snippet demonstrating how to use this
```

## Before your PR is "*Ready for review*"

- **Make sure you read and follow [Contributor guidelines](https://github.com/NVIDIA/Model-Optimizer/blob/main/CONTRIBUTING.md)** and your commits are signed.
- **Is this change backward compatible?**: Yes/No 
- **Did you write any new necessary tests?**: Yes/No
- **Did you add or update any necessary documentation?**: Yes/No
- **Did you update [Changelog](https://github.com/NVIDIA/Model-Optimizer/blob/main/CHANGELOG.rst)?**: Yes/No

### CODE REVIEW

**Comment by yeyu-nvidia** [comment]
/ok to test [cdea9ed](https://github.com/NVIDIA/Model-Optimizer/pull/664/commits/cdea9ed1450787d3164b1972fef72aed90c5665d)

**Comment by yeyu-nvidia** [comment]
/ok to test [f8a1088](https://github.com/NVIDIA/Model-Optimizer/pull/664/commits/f8a10886306092a960893d464d4aecee18285c84)

File: modelopt/torch/speculative/plugins/transformers.py [hunk 1]
```diff
@@ -456,6 +465,19 @@ def modify(
             eagle_architecture_config=eagle_architecture_config,
         )
         self.eagle_config = PretrainedConfig.from_dict(eagle_architecture_config)
+        # Hidden size and vocab size must match base model
+        llm_config = self.config.llm_config if hasattr(self.config, "llm_config") else self.config
```
Comment (h-guo18) [inline modelopt/torch/speculative/plugins/transformers.py]: nit: I think we can just use `self._base_llm_config` property which is defined [here](https://github.com/NVIDIA/Model-Optimizer/blob/2a7d7fe9e16d9402a31a779043fe52466d146286/modelopt/torch/speculative/plugins/transformers.py#L361)

File: modelopt/torch/speculative/plugins/transformers.py [hunk 2]
```diff
@@ -456,6 +465,19 @@ def modify(
...
+        self.eagle_config.hidden_size = llm_config.hidden_size
+        self.eagle_config.vocab_size = llm_config.vocab_size
+        self.eagle_config.max_position_embeddings = llm_config.max_position_embeddings
+        self.eagle_config.draft_vocab_size = getattr(
+            self.eagle_config, "draft_vocab_size", self.eagle_config.vocab_size
+        )
+        if getattr(self.eagle_config, "head_dim", None) is None:
```
Comment (h-guo18) [inline modelopt/torch/speculative/plugins/transformers.py]: Seems like this is an unreachable branch, since we have an default head_dim [here](https://github.com/NVIDIA/Model-Optimizer/blob/main/modelopt/torch/speculative/eagle/default_config.py#L51)?

File: modelopt/torch/speculative/plugins/transformers.py [hunk 3]
```diff
@@ -731,7 +746,14 @@ def _eagle_forward(
         )
         eagle_logits = eagle_lm_head(eagle_postnorm_h)
 
-        return eagle_postnorm_h, eagle_prenorm_h, eagle_logits, eagle_cache
+        draft_logits_list = [eagle_logits]
+        if self.eagle_config.parallel_draft_step > 1:
+            # Get additional draft logits from parallel draft heads
+            for draft_head in self.eagle_module.parallel_draft_heads:
```
Comment (h-guo18) [inline modelopt/torch/speculative/plugins/transformers.py:752]: I think we can optimize this for-loop and do it parallelly. Can be done in following PRs perhaps.

File: modelopt/torch/speculative/plugins/transformers.py [hunk 4]
```diff
@@ -855,24 +875,31 @@ def forward(
...
-                # base model predict +1 tok, while eagle predict +2
-                # so we shift base model outputs compared to eagle outputs
-                base_model_logits[:, 1:],
-                eagle_logits[:, :-1],
-                # additionally, we mask the first n tok of eagle outputs at nth TTT step
-                torch.cat(
-                    (
-                        torch.zeros(b, ttt_step, dtype=loss_mask.dtype, device=loss_mask.device),
-                        loss_mask[:, 1 + ttt_step :],
+            for i in range(self.eagle_config.parallel_draft_step):
```
Comment (h-guo18) [inline modelopt/torch/speculative/plugins/transformers.py:878]: Similar as above, this for-loop seems parallelizable

File: modelopt/torch/speculative/plugins/transformers.py
Comment (h-guo18) [inline modelopt/torch/speculative/plugins/transformers.py]: The semantic of this `steps` argument seems ambiguous to me. Shall we rename it to `eagle_steps`? Then it means we do `eagle_steps` sequential drafting + `num_medusa_heads` parallel drafrting.

File: modelopt/torch/speculative/plugins/transformers.py [hunk 5]
```diff
@@ -987,7 +1017,12 @@ def pseudo_speculative_generate(
                     position_embeddings,
                 )
 
-            draft_token = eagle_logits[:, -1:, :].argmax(dim=-1)
+            if self.eagle_config.parallel_draft_step > 1:
+                parallel_logits = [
```
Comment (h-guo18) [inline modelopt/torch/speculative/plugins/transformers.py:1022]: Seems like only the last eagle step's parallel logits are used? Shall we only compute this in the last eagle step?

File: modelopt/torch/speculative/plugins/transformers.py (hunk 2 above)
Comment (yeyu-nvidia) [reply modelopt/torch/speculative/plugins/transformers.py]: removed

File: examples/speculative_decoding/eagle_config.json [hunk 6]
```diff
@@ -1,11 +1,3 @@
 {
-    "rope_scaling": {
```
Comment (sugunav14) [inline examples/speculative_decoding/eagle_config.json:2]: why do we need to delete this configuration?

Comment (yeyu-nvidia) [reply examples/speculative_decoding/eagle_config.json:2]: It's duplicated. We already have this in the default config here https://github.com/NVIDIA/Model-Optimizer/blob/main/modelopt/torch/speculative/eagle/default_config.py

**Review by sugunav14 — APPROVED** [review]

File: modelopt/torch/export/plugins/hf_spec_export.py [hunk 7]
```diff
@@ -18,26 +18,62 @@
...
+        eagle_modelopt_to_official["required"].update(
+            {
+                "hidden_norm.weight": "midlayer.hidden_norm.weight",
+                "input_embeds_norm.weight": "midlayer.input_layernorm.weight",
+            }
+        )
+    else:
+        eagle_modelopt_to_official["required"].update(
+            {
+                "hidden_norm.weight": "layers.0.hidden_norm.weight",
```
Comment (h-guo18) [inline modelopt/torch/export/plugins/hf_spec_export.py]: Why are these two params naming needs to be different for single and multi layer eagle? Shall we make them the same for conciseness?

Comment (yeyu-nvidia) [reply modelopt/torch/export/plugins/hf_spec_export.py]: Typo. Fixed
